### DISCUSSION TITLE
Docker-OSX for iOS CI/CD

### DISCUSSION THREAD

**Discussion by hisaac** [MobileNativeFoundation/discussions#63] [General]
Is anyone using the [Docker-OSX](https://github.com/sickcodes/Docker-OSX) project in an iOS CI/CD pipeline? I'd love to hear your experiences with it.

**Comment by Malinskiy** [comment]
I've implemented an internal company-wide system with docker containers running virtualised macOS instances ~4 years ago when there wasn't a separate project like Docker-OSX on the market.

Although I can't speak to the quality of the Docker-OSX, I can in general provide feedback on the approach:
The pros:
* Reusing existing Kubernetes clusters for macOS orchestration (unified monitoring of instance, all the metrics you get out of the box are all there). Just taint the nodes and scale on demand how many macOS environments you need
* Reproducibility: you can quickly get to a known state of the 
* Negligible (<10%) performance impact in benchmarks and no measurable performance impact when executing real CI tasks, e.g. builds/tests

Cons:
* Requires KVM over IP (the physical KVM, not the linux virtualisation) to install and maintain the Linux for the network boot/remote maintenance or recovery. Not all KVM solutions support Apple hardware smoothly, on top of this there might be a need for real physical power supply separate to the Apple's device (in my particular case the USB power for the KVM virtual keyboard came up after the boot checks if it needs to boot from network/recovery). KVM over IP is a con though of every solution with Apple hardware: Apple doesn't provide remote management. The only difference is sometimes you own it (like in this case) or in some cases another team owns it.
* Native network ethernet adapters on Apple's hardware are notoriously buggy in Linux kernel during sustained workload, so realistically you will need a secondary ethernet card bonded into a network interface (active/passive)
* Need to keep the Apple hardware taints in the kube cluster so that you still abide by the legal requirement of running macOS on Apple's hardware

Notes:
* Differential disks and snapshots really help with spawning a new macOS instance
* T2 chip requires the use of external storage but doesn't stop you from doing anything else

The experience has been very positive and I would prefer this solution over virtual machines running in macOS using xhyve or similar. The main reason for me was consistency across the tools for monitoring health and the uptime of the infrastructure. I'm not sure this will work though with the latest M1 hardware, but since there are already efforts put into running Linux on M1, it's a matter of time before KVM will provide the same solution for M1 macs.

Hope this helps

**Reply by hisaac** [reply]
Thank you for the detailed information @Malinskiy!

You mentioned at the beginning that you implemented docker containers running macOS in the past before this Docker-OSX project. I'm not very familiar with Docker yet, so forgive my ignorance, but I thought that Docker didn't natively support macOS for some reason. How did you implement it?

**Reply by Malinskiy** [reply]
The non-native support of docker is when you run docker _inside_ the macOS.
What I was doing and what Docker-OSX is doing at the moment is running macOS inside a container.

It really is just a virtual machine with the runtime of it packaged into a container. The same is used now for running the Android Emulator in docker: passthrough the /dev/kvm or support nested virtualisation and voila, you're using a qemu distribution that you package via docker and connect it to the local KVM in kernel.

**Reply by hisaac** [reply]
Ah, so Docker can't run a macOS VM on a macOS host machine, but it can run a macOS VM on a Linux host machine? Is that right?

**Reply by Malinskiy** [reply]
Yes, you basically install Linux natively on Apple hardware and then use that to boot up virtualised macOS

**Reply by hisaac** [reply]
Gotcha, that's great information. Thank you! 🙏
